search: Add filter section for direct messages.

lakshaykananiya opened the pull request:
This PR adds the feature of filtering DMs just like filtering topics in streams.

Fixes: https://github.com/zulip/zulip/issues/22113

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![filter](https://user-images.githubusercontent.com/101464851/225063889-f21b36ee-fc0e-4c01-9ac7-9f2d93e06c96.gif)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks!

1. There shouldn't be a "more conversations" link at the bottom. The use is already in that view. Also, clicking on it causes crazy things to happen. ;)

![Screen Shot 2023-03-14 at 10 09 22 AM](https://user-images.githubusercontent.com/2090066/225084301-5bbbb51a-c6be-44cd-b56f-a4718aef8d73.png)

2. Let's change the placeholder text to "Filter direct messages" -- I think there's plenty of room for it.

alya commented:
Also, I think it would feel nicer if the UI offered typeahead suggestions for names, same as in the "To" line of the compose box, but ideally filtered down to the users with whom you have PM conversations.

lakshaykananiya commented:
> Also, I think it would feel nicer if the UI offered typeahead suggestions for names, same as in the "To" line of the compose box, but ideally filtered down to the users with whom you have PM conversations.

Isn't it the same thing filter is doing?

alya commented:
> Isn't it the same thing filter is doing?

Hm, yeah, maybe we don't need it.

lakshaykananiya commented:
Thanks, fixed the background issue. It now looks something like this,

<img width="263" alt="light" src="https://user-images.githubusercontent.com/101464851/229122762-ab4c2520-061e-41c7-961d-9e3b3441267a.png"><img width="264" alt="dark" src="https://user-images.githubusercontent.com/101464851/229122777-c8757b0a-4fb6-46fc-8e6f-add153de8173.png">

@amanagr, FYI.

alya commented:
Looks like the next step here is resolving the conflicts.

amanagr commented on web/src/pm_list.js:
```diff
@@ -59,6 +61,19 @@ export function _build_private_messages_list() {
+    const $input = $("#filter-pm-input");
+    if ($input.length) {
+        $input.val("");
+        $input.trigger("blur");
```
I would have preferred if the focused stayed on the input after clicking on the clear button since I am not trying to close the expanded pm list by clicking on the close button but most likely trying to search for another user.

lakshaykananiya replied:
Hmm, looks like we're not maintaining focus on any of the filter or search bar when we clear it by clicking on the cross right now. Is it because if someone wants to continue the search they may not want to touch the mouse and click on the icon and rather use `cmd + a`to select everything and type the new query? It still feels like it would be good to have it focussed anyway.

lakshaykananiya replied:
Updated!

amanagr replied:
> Hmm, looks like we're not maintaining focus on any of the filter or search bar when we clear it by clicking on the cross right now. 

Oh! You can add a commit to fix that in this PR too.

amanagr replied:
or, you can start a frontend discussion for it and open a new PR based on the discussion.

amanagr commented on web/src/pm_list.js:
```diff
@@ -194,6 +215,9 @@ function zoom_in() {
     $("#streams_list").hide();
     $(".left-sidebar .right-sidebar-items").hide();
+
+    // Add filter section for direct messages.
+    $(".private_messages_container").children("div.pm_filter_row").append(render_filter_pms());
```
I don't think we need the `children` query here, we can do this in one single query.

lakshaykananiya replied:
Updated!

amanagr commented on web/src/pm_list.js:
```diff
@@ -218,4 +253,8 @@ export function initialize() {
     });
+
+    $("body").on("input", "#filter-pm-input", () => {
+        update_private_messages();
```
Let's add `e.stopPropagation();` and `e.preventDefault();` here.

lakshaykananiya replied:
Updated!

amanagr commented:
@lakshaykananiya thanks for working on this! Left some comments.

lakshaykananiya commented:
@amanagr I've addressed all your feedback. It now does not lose focus when clicking on the cross, also I've updated the code with the other two changes you suggested.

lakshaykananiya commented:
Added commit to maintain focus on other filter bars too.
![a](https://user-images.githubusercontent.com/101464851/230338274-26937b80-463d-4b46-a019-6f3727a5b11a.gif)
![b](https://user-images.githubusercontent.com/101464851/230338289-8e458916-dd2e-4cd8-9055-e75522a5bbfe.gif)

amanagr commented:
LGTM

timabbott commented on web/src/pm_list_data.js:
```diff
@@ -81,9 +83,17 @@ export function get_conversations() {
 
     if (zoomed || conversations.length <= max_conversations_to_show) {
+        if (zoomed) {
+            const search_term = pm_list.get_pm_search_term();
```
The caller should pass the `search_term` in to avoid an import cycle; we just fixed this in the stream list version of this code.

timabbott commented on web/src/pm_list.js:
```diff
@@ -202,6 +226,17 @@ function zoom_out() {
     $(".left-sidebar .right-sidebar-items").show();
+
+    // Remove filter section for direct messages.
+    $(".filter-pms").remove();
```
Can we write this as `$(".pm_filter_row .filter-pms").remove();` for closer symmetry with the above?

Even better would be something like `$(".pm_filter_row").empty();`

lakshaykananiya replied:
Updated!

timabbott commented on web/styles/app_components.css:
```diff
@@ -466,9 +466,17 @@ div.overlay {
+        width: calc(100% - 36px);
+        margin-left: 11px;
+        margin-bottom: 5px;
+    }
```
Reading this code -- is it really the case that `stream-list-filter` has 1px different margin than these others?

We can leave this for a follow-up, but it seems likely we should aim to rewrite these filter elements as a new `sidebar-filter-input` component.

lakshaykananiya replied:
Though it should have been but looks like it is not applicable at all on `stream-list-filter` and rather another property of same kind from `left-sidebar.css` is applicable.

For `pm-list-filter` this was actually a bug and there was no need to consider margin at all, I've fixed that!

lakshaykananiya replied:
The reason behind the 11px could be that there is an extra `margin-left` of `3px` along with a `padding-left` of `20px` which it is trying to adjust as 11px on both sides (hence x2). I am not sure though why was there that extra margin.

lakshaykananiya commented:
@timabbott Mentioning for a quick re-review!

Rinwaoluwa commented:
@lakshaykananiya Are you still working on this ? I would like to pick up from where you left off, It's fine if you still are I'll take up something else.

lakshaykananiya commented:
@Rinwaoluwa You are welcome to pick this up, I'm not working on it anymore.

Rinwaoluwa commented:
@lakshaykananiya Thank you I'll be sure to add you to the co author
